Extending the CCDB API to optionally get the headers from first HTTP request

mrtineide opened the pull request:
### Some background
There is [some](https://github.com/AliceO2Group/O2Physics/blob/b11abd5ee846354188fc1bba76fed8fa48ca588f/Common/TableProducer/PID/pidTPCModule.h#L341) [code](https://github.com/AliceO2Group/O2Physics/blob/b11abd5ee846354188fc1bba76fed8fa48ca588f/Common/TableProducer/PID/pidTPC.cxx#L243) that used the retrieveHeaders function right after getting a object from CCDB, sending one request for the object itself and then one HTTP HEAD  request immediately after for the metadata/headers.

I was asked by @costing to change the O2/CCDB C++ API such we could return the headers in one call, enabling the use of just one call to CCDB, avoiding the extra network calls/retrieveHeaders. 

This PR implements these changes.

##### New functionality
- Extended the CcdbApi::retrieveBlob to take a optional header paramater to return metadata/headers 
- Extend the BasicCCDBmanager to cache and return headers/metadata if asked for

This does increase the memory used,  but I guess it is negligible. 

(I also added some unit tests to make myself feel more comfortable with my changes)

mrtineide committed:
Extend CCDB functions with optional headers

mrtineide committed:
Remove the header filtering & add more tests

mrtineide committed:
Fix code snippets and enhance metadata retrieval documentation

Updated code snippets in README for header retrieval and added optional parameter documentation details.
Also fixed some CI formatting errors.

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 9f2938baea2b03613f0a298e537904d8d88423ce at 2025-10-01 12:56:
```
## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/SOURCES/O2/14709-slc9_x86-64/0/CCDB/test/testCcdbApiHeaders.cxx:43:42: error: statement should be inside braces [readability-braces-around-statements]
++ [[ 0 == 0 ]]
++ exit 1
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14709/9f2938baea2b03613f0a298e537904d8d88423ce/build_O2_fullCI_slc9/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI_slc9 for d7b08a5e009bf20c77fa459eea07cd74e52fed12 at 2025-10-01 16:13:
```
++ cp /sw/slc9_x86-64/O2/14709-slc9_x86-64-local1/compile_commands.json .
+++ python3 -c 'import json, os; print(os.path.commonpath([x["file"] for x in json.loads(open("compile_commands.json").read()) if "sw/BUILD" not in x["file"] and "G__" not in x["file"] and x["file"].endswith(".cxx")]))'
++ O2_SRC=/sw/SOURCES/O2/14709-slc9_x86-64/0
++ [[ -e /sw/SOURCES/O2/14709-slc9_x86-64/0/CMakeLists.txt ]]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14709/d7b08a5e009bf20c77fa459eea07cd74e52fed12/build_O2_fullCI_slc9/pretty.html).

mrtineide commented:
I see some of the unit test that I made are flaky, I am able to get tests locally to fail randomly when running them many times. 
I think it has to do with some race condition or similar parallel execution regarding BasicCCDBManager singleton.

I'll change/remove the test are being flaky, so they do not pop up in other unrelated CI runs. I believe it is unrelated to the changes in this PR.

sawenzel commented:
> I see some of the unit test that I made are flaky, I am able to get tests locally to fail randomly when running them many times. I think it has to do with some race condition or similar parallel execution regarding BasicCCDBManager singleton.
> 
> I'll change/remove the test are being flaky, so they do not pop up in other unrelated CI runs. I believe it is unrelated to the changes in this PR.

I do not see how in a unit test there should be parallel execution and race conditions with BasicCCDBManager unless you talk to it in multiple threads.

sawenzel commented on CCDB/test/testCcdbApiHeaders.cxx:
```diff
@@ -0,0 +1,429 @@
+    hostReachable = api.isHostReachable();
+    char hostname[_POSIX_HOST_NAME_MAX];
+    gethostname(hostname, _POSIX_HOST_NAME_MAX);
+    basePath = std::string("Users/m/meide/BasicCCDBManager/");
```
paths like this should be avoided

sawenzel commented on CCDB/test/testCcdbApiHeaders.cxx:
```diff
@@ -0,0 +1,429 @@
+  f.api.storeAsTFileAny<std::string>(&ccdbObjN, pathB, md, start, stop);
+  f.api.storeAsTFileAny<std::string>(&ccdbObjX, pathC, md, start, stop);
+  // initilize the BasicCCDBManager
+  o2::ccdb::BasicCCDBManager& ccdbManager = o2::ccdb::BasicCCDBManager::instance();
```
prefer using `auto& mgr = ` ... like you do further below

mrtineide replied:
[Agree, fixed now](https://github.com/AliceO2Group/AliceO2/pull/14709/commits/2246830d98e57d749247e155e2b2af66f5e29d3c#diff-c4759010f03b9948d6efde2e67efb3c66baae9c3bf750340932cfd359def9009R176-L207).

mrtineide committed:
Change unit test with to have correct includes

This fixes some tests that are fine on my desktop but
the subprocess aborts even when all tests are green in the CI.
The error is this: malloc_consolidate(): invalid chunk size
Was told this was a heap corruption. And changing the
includes fixes the error, not sure why.
Changed also to only use one instance of CCDB api
so we do not call CURL global cleanup out of order
with the BasicCCDBManger singelton. Just in case
that matters when running the test(s).

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 60ede12ba4eed3139a8a31350a1c3534823de179 at 2025-10-08 03:40:
```
No log files found
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14709/60ede12ba4eed3139a8a31350a1c3534823de179/build_O2_fullCI_slc9/pretty.html).

mrtineide commented:
> > I see some of the unit test that I made are flaky, I am able to get tests locally to fail randomly when running them many times. I think it has to do with some race condition or similar parallel execution regarding BasicCCDBManager singleton.
> > I'll change/remove the test are being flaky, so they do not pop up in other unrelated CI runs. I believe it is unrelated to the changes in this PR.
> 
> I do not see how in a unit test there should be parallel execution and race conditions with BasicCCDBManager unless you talk to it in multiple threads.

That was what I initially though was happening, but I was incorrect. But turns out I included the wrong header in the unit test. I should have used `#include <boost/test/unit_test.hpp` and not `#include <boost/test/included/unit_test.hpp`!

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 2246830d98e57d749247e155e2b2af66f5e29d3c at 2025-10-28 20:20:
```
## sw/BUILD/O2Physics-latest/log
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14709/2246830d98e57d749247e155e2b2af66f5e29d3c/build_O2_fullCI_slc9/pretty.html).

jezwilkinson commented:
Hi, is there any news about the review of this PR? It would be good to have the API changes in place so that we can close https://github.com/AliceO2Group/O2Physics/pull/13265
